pkg/drivers/docker_driver.go

```diff
@@ -0,0 +1,371 @@
+		originalImage: args.Image,
+		currentImage:  args.Image,
+		cli:           *newCli,
+		env:           nil,
+		save:          args.Save,
+	}, nil
+}
+
+func (d *DockerDriver) Destroy() {
+	// since intermediate images are chained, removing the most current
+	// image (that isn't the original) removes all previous ones as well.
+	if d.currentImage != d.originalImage {
+		if err := d.cli.RemoveImage(d.currentImage); err != nil {
+			logrus.Infof("error removing image: %s", err)
+		}
+	}
+}
+
+func (d *DockerDriver) Setup(envVars []unversioned.EnvVar, fullCommands [][]string) {
+	env := d.processEnvVars(envVars)
+	for _, cmd := range fullCommands {
+		d.currentImage = d.runAndCommit(env, cmd)
+	}
+}
+
+func (d *DockerDriver) Teardown(envVars []unversioned.EnvVar, fullCommands [][]string) {
+	// since we create a new driver for each test, skip teardown commands
+	logrus.Debug("Docker driver does not support teardown commands, since each test gets a new driver. Skipping commands.")
+}
+
+func (d *DockerDriver) ProcessCommand(envVars []unversioned.EnvVar, fullCommand []string) (string, string, int) {
+	var env []string
+	for _, envVar := range envVars {
+		env = append(env, envVar.Key+"="+envVar.Value)
+	}
+	stdout, stderr, exitCode := d.exec(env, fullCommand)
+
+	if stdout != "" {
+		logrus.Infof("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		logrus.Infof("stderr: %s", stderr)
+	}
+	return stdout, stderr, exitCode
+}
+
+func retrieveEnv(d *DockerDriver) func(string) string {
+	return func(envVar string) string {
+		var env map[string]string
+		if env == nil {
+			image, err := d.cli.InspectImage(d.currentImage)
+			if err != nil {
+				return ""
+			}
+			// convert env to map for processing
+			env = convertEnvToMap(image.Config.Env)
+		}
+		return env[envVar]
+	}
+}
+
+func (d *DockerDriver) retrieveEnvVar(envVar string) string {
+	// since we're only retrieving these during processing, we can use a closure to cache this
+	return retrieveEnv(d)(envVar)
+}
+
+func (d *DockerDriver) processEnvVars(vars []unversioned.EnvVar) []string {
+	if len(vars) == 0 {
+		return nil
+	}
+
+	env := []string{}
+
+	for _, envVar := range vars {
+		expandedVal := os.Expand(envVar.Value, d.retrieveEnvVar)
+		env = append(env, envVar.Key+"="+expandedVal)
+	}
+	return env
+}
+
+// copies a tar archive starting at the specified path from the image, and returns
+// a tar reader which can be used to iterate through its contents and retrieve metadata
+func (d *DockerDriver) retrieveTar(path string) (*tar.Reader, error) {
+	// this contains a placeholder command which does not get run, since
+	// the client doesn't allow creating a container without a command.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image: d.currentImage,
+			Cmd:   []string{"NOOP_COMMAND_DO_NOT_RUN"},
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+	})
+	if err != nil {
+		logrus.Errorf("Error creating container: %s", err.Error())
```

Comment:
I don't think you want to log the error here, you can use an `errors.Wrapf` to add context to the error but it should be handled at top level most likely.

Reply:
reworked the error handling here.